### PR TITLE
Adding support for Copr-build job in gitlab

### PR DESCRIPTION
**Pull request by shreyaspapi**
- [x] Tests

### CODE REVIEW

File: packit_service/config.py [hunk 1]
```diff
@@ -65,6 +65,7 @@ def __init__(
         bugzilla_url: str = "",
         bugzilla_api_key: str = "",
         pr_accepted_labels: List[str] = None,
+        gitlab_webhook_token: str = "",
```
Comment (lachmanfrantisek) [inline packit_service/config.py:68]: I would pick a list of allowed token's if it does not complicate anything.

File: packit_service/service/events.py [hunk 2]
```diff
@@ -485,7 +485,7 @@ def __init__(
         self.target_repo_namespace = target_repo_namespace
         self.target_repo_name = target_repo_name
         self.https_url = https_url
-        self.username = username
+        self.user_login = username
```
Comment (lachmanfrantisek) [inline packit_service/service/events.py:488]: Why this change? (I just want to know.)

File: packit_service/worker/handlers/gitlab_handlers.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+logger = logging.getLogger(__name__)
+
+
+@add_to_comment_action_mapping
+@add_to_comment_action_mapping_with_name(name=CommentAction.build)
+@use_for(JobType.build)
+@use_for(JobType.copr_build)
+@required_by(JobType.tests)
+class GitlabMergeRequestCommentCoprBuildHandler(CommentActionHandler):
+    """ Handler for MR comment `/packit copr-build` """
```
Comment (lachmanfrantisek) [inline packit_service/worker/handlers/gitlab_handlers.py:56]: Is this really needed? Cannot we reuse Github one?

Comment (shreyaspapi) [reply packit_service/worker/handlers/gitlab_handlers.py:56]: Yes, we can I made some mistake with git. This as not supposed to be here.

File: packit_service/config.py (hunk 1 above)
Comment (shreyaspapi) [reply packit_service/config.py:68]: Sure.

File: packit_service/service/events.py (hunk 2 above)
Comment (shreyaspapi) [reply packit_service/service/events.py:488]: To reuse `GitHubPullRequestCommentCoprBuildHandler` as it should have `self.event.user_login`. So just renaming username -> user_login.

Comment (lachmanfrantisek) [reply packit_service/service/events.py:488]: Thanks!

**Comment by shreyaspapi** [comment]
Fixing the git issue.

File: packit_service/config.py (hunk 1 above)
Comment (shreyaspapi) [reply packit_service/config.py]: Done made it into the list!

Comment (lachmanfrantisek) [reply packit_service/config.py]: Thanks!

File: packit_service/config.py [hunk 4]
```diff
@@ -65,6 +65,7 @@ def __init__(
         bugzilla_url: str = "",
         bugzilla_api_key: str = "",
         pr_accepted_labels: List[str] = None,
+        gitlab_webhook_token: List[str] = [],
```
Comment (lachmanfrantisek) [inline packit_service/config.py:68]:
```suggestion
        gitlab_webhook_tokens: List[str] = [],
```

File: packit_service/config.py [hunk 5]
```diff
@@ -89,6 +90,9 @@ def __init__(
         # for flask SERVER_NAME so we can create links to logs
         self.server_name: str = ""
 
+        # Makeshift for now to authenticate webhooks coming from gitlab instances
+        self.gitlab_webhook_token: Set[str] = set(gitlab_webhook_token or [])
```
Comment (lachmanfrantisek) [inline packit_service/config.py:94]:
```suggestion
        self.gitlab_webhook_tokens: Set[str] = set(gitlab_webhook_tokens or [])
```

File: packit_service/config.py [hunk 6]
```diff
@@ -104,6 +108,7 @@ def hide(token: str) -> str:
             f"fas_password='{hide(self.fas_password)}', "
             f"bugzilla_url='{self.bugzilla_url}', "
             f"bugzilla_api_key='{hide(self.bugzilla_api_key)}', "
+            f"gitlab_webhook_token='{self.gitlab_webhook_token}',"
```
Comment (lachmanfrantisek) [inline packit_service/config.py:111]:
```suggestion
            f"gitlab_webhook_tokens='{self.gitlab_webhook_tokens}',"
```

File: packit_service/schema.py [hunk 7]
```diff
@@ -56,6 +56,7 @@ class ServiceConfigSchema(UserConfigSchema):
     pr_accepted_labels = fields.List(fields.String())
     admins = fields.List(fields.String())
     server_name = fields.String()
+    gitlab_webhook_token = fields.List(fields.String())
```
Comment (lachmanfrantisek) [inline packit_service/schema.py:59]:
```suggestion
    gitlab_webhook_tokens = fields.List(fields.String())
```

File: packit_service/service/api/webhooks.py [hunk 8]
```diff
@@ -224,7 +224,7 @@ def validate_token():
         token = request.headers["X-Gitlab-Token"]
 
         # Find a better solution
-        if token != config.gitlab_webhook_token:
+        if token not in config.gitlab_webhook_token:
```
Comment (lachmanfrantisek) [inline packit_service/service/api/webhooks.py:227]:
```suggestion
        if token not in config.gitlab_webhook_tokens:
```

File: packit_service/worker/handlers/github_handlers.py [hunk 9]
```diff
@@ -505,25 +522,39 @@ class GitHubPullRequestCommentCoprBuildHandler(CommentActionHandler):
...
-        user_can_merge_pr = self.event.project.can_merge_pr(self.event.user_login)
-        if not (user_can_merge_pr or self.event.user_login in self.config.admins):
-            self.event.project.pr_comment(
-                self.event.pr_id, PERMISSIONS_ERROR_WRITE_OR_ADMIN
-            )
-            return HandlerResults(
-                success=True, details={"msg": PERMISSIONS_ERROR_WRITE_OR_ADMIN}
+        if isinstance(
+            self.event, (PullRequestCommentGithubEvent, MergeRequestCommentGitlabEvent)
+        ):
```
Comment (lachmanfrantisek) [inline packit_service/worker/handlers/github_handlers.py:527]: What is the `else` variant here? Can we add some comment about that? (It will go silently without permissions check so we might want to be more explicit here.)

File: tests/unit/test_config.py [hunk 10]
```diff
@@ -62,6 +62,7 @@ def service_config_valid():
         "command_handler_k8s_namespace": "packit-test-sandbox",
         "admins": ["Dasher", "Dancer", "Vixen", "Comet", "Blitzen"],
         "server_name": "hub.packit.org",
+        "gitlab_webhook_token": ["token1", "token2", "token3", "aged"],
```
Comment (lachmanfrantisek) [inline tests/unit/test_config.py:65]:
```suggestion
        "gitlab_webhook_tokens": ["token1", "token2", "token3", "aged"],
```

File: tests/unit/test_config.py [hunk 11]
```diff
@@ -81,6 +82,7 @@ def test_parse_valid(service_config_valid):
     assert config.command_handler_work_dir == "/sandcastle"
     assert config.admins == {"Dasher", "Dancer", "Vixen", "Comet", "Blitzen"}
     assert config.server_name == "hub.packit.org"
+    assert config.gitlab_webhook_token == {"token1", "token2", "token3", "aged"}
```
Comment (lachmanfrantisek) [inline tests/unit/test_config.py:85]:
```suggestion
    assert config.gitlab_webhook_tokens == {"token1", "token2", "token3", "aged"}
```

File: tests/unit/test_config.py [hunk 12]
```diff
@@ -133,6 +135,7 @@ def test_config_opts(sc):
     assert sc.github_requests_log_path is not None
     assert sc.webhook_secret is not None
     assert sc.validate_webhooks is not None
+    assert sc.gitlab_webhook_token is not None
```
Comment (lachmanfrantisek) [inline tests/unit/test_config.py:138]:
```suggestion
    assert sc.gitlab_webhook_tokens is not None
```

File: packit_service/worker/handlers/github_handlers.py (hunk 9 above)
Comment (shreyaspapi) [reply packit_service/worker/handlers/github_handlers.py]: Yeah! I have removed that check as we are already checking the events in pre-check.

**Comment by shreyaspapi** [comment]
Thanks for the review.

Rebased!

**Comment by lachmanfrantisek** [comment]
Tried tests locally and one failed:

```
tests_requre/database/test_models.py:110 (test_get_merged_chroots)
fedora-43-x86_64 != fedora-42-x86_64

Expected :fedora-42-x86_64
Actual   :fedora-43-x86_64
<Click to see difference>

clean_before_and_after = None, too_many_copr_builds = None

    def test_get_merged_chroots(clean_before_and_after, too_many_copr_builds):
        # fetch 10 merged groups of builds
        builds_list = list(CoprBuildModel.get_merged_chroots(10, 20))
        assert len(builds_list) == 10
        # two merged chroots so two statuses
        assert len(builds_list[0].status) == 2
        assert len(builds_list[0].target) == 2
        assert builds_list[1].status[0][0] == "success"
>       assert builds_list[2].target[0][0] == "fedora-42-x86_64"
E       AssertionError: assert 'fedora-43-x86_64' == 'fedora-42-x86_64'

database/test_models.py:119: AssertionError
```

Looks like it is not sorted.

**Comment by lachmanfrantisek** [comment]
Probably some fault on my side. The failing test is not new and passed in CI in the past.

**Comment by shreyaspapi** [comment]
I am unable to recreate this issue.

**Comment by IceWreck** [comment]
@lachmanfrantisek actually, its fails randomly for me too. i'll look into it

**Comment by shreyaspapi** [comment]
This took me 20-30 tries to replicate.

Having it here just for reference.

```
clean_before_and_after = None, too_many_copr_builds = None

    def test_get_merged_chroots(clean_before_and_after, too_many_copr_builds):
        # fetch 10 merged groups of builds
        builds_list = list(CoprBuildModel.get_merged_chroots(10, 20))
        assert len(builds_list) == 10
        # two merged chroots so two statuses
        assert len(builds_list[0].status) == 2
        assert len(builds_list[0].target) == 2
>       assert builds_list[1].status[0][0] == "success"
E       AssertionError: assert 'pending' == 'success'
E         - success
E         + pending
```
